### PR TITLE
Fix implementation of hd derivation to be bip32 compliant

### PR DESCRIPTION
**Pull request by braydonf**
https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#child-key-derivation-ckd-functions

Closes: https://github.com/bitpay/bitcore-lib/issues/94
Closes: https://github.com/bitpay/bitcore-lib/issues/93

### CODE REVIEW

**Comment by axic** [comment]
Great work! Can you also include #93?

**Comment by braydonf** [comment]
There is validation of the private key at: https://github.com/bitpay/bitcore-lib/blob/master/lib/hdprivatekey.js#L416 when the private key is constructed, however I don't see logic for proceeding to the next value for i.

**Comment by braydonf** [comment]
Okay, handling of invalid private key derivation is handled. The test was broken before, and errored with a message about an invalid private key rather than incrementing the index.

**Comment by braydonf** [comment]
There may also be this case: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#public-parent-key--public-child-key

**Comment by braydonf** [comment]
Check the bitcore-lib tests against libbtc here: https://github.com/braydonf/libbtc/tree/cross-ref There seems to be an issue with the fingerprint in the test that still needs to be worked out.

**Comment by flyswatter** [comment]
When can we look forward to having this merged & pushed to npm? There are a lot of broken HD wallets out there right now, would be nice to get people using the same algorithms.

**Comment by gabegattis** [comment]
@flyswatter Hopefully, we can get our 1.0.0 release (which should include this fix) sometime in the next couple of weeks.

**Comment by braydonf** [comment]
Also tested against hdkey at https://github.com/cryptocoinjs/hdkey/pull/9

**Comment by braydonf** [comment]
And with `bcoin` as well at https://github.com/bcoin-org/bcoin/pull/84

**Comment by braydonf** [comment]
This PR has been updated from feedback in https://github.com/bitpay/bitcore-lib/pull/115

**Comment by matiu** [comment]
ACK

**Comment by gabegattis** [comment]
The plan was to leave .derive() the way it is for the time being.

The 2 new methods would be be for getting a compliant or non-compliant child. We don't want to make .derive() throw util we make breaking changes in v1.0.0.

Also something I mentioned internally, but not on GitHub, I think we should remove the hdkey cache now. We know we want to remove it in v1.0.0. This PR requires changes to the cache code. We can avoid that by just removing the cache. Removing the cache would not be a breaking change. It is not documented and, where it is exposed in index.js, we comment that it is only exposed for internal usage and testing.

**Comment by gabegattis** [comment]
I opened a new PR that reflects my comment above.
https://github.com/bitpay/bitcore-lib/pull/116

**Comment by braydonf** [comment]
Removing hdkeycache can be it's own atomic change and is not related to fixing BIP32 derivation.

File: lib/hdprivatekey.js [hunk 1]
```diff
@@ -139,28 +144,51 @@ HDPrivateKey._getDerivationIndexes = function(path) {
...
-HDPrivateKey.prototype._deriveWithNumber = function(index, hardened) {
+/**
+ * WARNING: If this is a new implementation you should NOT use this method, you should be using
+ * `derive` instead.
+ *
+ * This method is explicitly for use and compatibility with an implementation that
+ * was not compliant with BIP32 regarding the derivation algorithm. The private key
+ * must be 32 bytes hashing, and this implementation will use the non-zero padded
+ * serialization of a private key, such that it's still possible to derive the privateKey
+ * to recover those funds.
```
Comment (Levino) [inline lib/hdprivatekey.js]: As much as I like this comment, it is easy to overlook it. One should log a warning when someone is using this function too.

File: lib/hdprivatekey.js [hunk 2]
```diff
@@ -172,15 +200,23 @@ HDPrivateKey.prototype._deriveWithNumber = function(index, hardened) {
     index += HDPrivateKey.Hardened;
   }
 
-  var cached = HDKeyCache.get(this.xprivkey, index, hardened);
+  var cached = HDKeyCache.get(this.xprivkey, index, hardened, nonCompliant);
```
Comment (Levino) [inline lib/hdprivatekey.js]: Same here. Do we really want to drag this "nonCompliant" parameter through the ages from now on?

File: lib/hdprivatekey.js [hunk 3]
```diff
@@ -172,15 +200,23 @@ HDPrivateKey.prototype._deriveWithNumber = function(index, hardened) {
...
   if (cached) {
     return cached;
   }
 
   var indexBuffer = BufferUtil.integerAsBuffer(index);
   var data;
-  if (hardened) {
-    data = BufferUtil.concat([new buffer.Buffer([0]), this.privateKey.toBuffer(), indexBuffer]);
+  if (hardened && nonCompliant) {
```
Comment (Levino) [inline lib/hdprivatekey.js]: What if it is ```hardened = false && nonCompliant = true```?

File: test/hdkeys.js [hunk 4]
```diff
@@ -158,69 +159,173 @@ describe('BIP32 compliance', function() {
...
+      childIndex: 0,
+      privateKey: privateKeyBuffer,
+      chainCode: chainCodeBuffer
+    });
+    var derived = key.deriveChild("m/44'/0'/0'/0/0'");
+    derived.privateKey.toString().should.equal('3348069561d2a0fb925e74bf198762acc47dce7db27372257d2d959a9e6f8aeb');
+  });
+
+  it('should NOT use full 32 bytes for private key data that is hashed with nonCompliant flag', function() {
+    // This is to test that the previously implemented non-compliant to BIP32
```
Comment (Levino) [inline test/hdkeys.js]: This is not a proper sentence...

File: test/hdkeys.js [hunk 5]
```diff
@@ -158,69 +159,173 @@ describe('BIP32 compliance', function() {
...
+      pubkey.derive();
+    }).to.throw('Method has been deprecated');
+  });
+
+  describe('edge cases', function() {
+    var sandbox = sinon.sandbox.create();
+    afterEach(function() {
+      sandbox.restore();
+    });
+    it('will handle edge case that derived private key is invalid', function() {
```
Comment (Levino) [inline test/hdkeys.js]: I do not understand what this test is about. Is it feeding in data that triggered the bug and now we have test that the bug is fixed? Maybe say so in the comments?

File: test/privatekey.js [hunk 6]
```diff
@@ -331,6 +331,20 @@ describe('PrivateKey', function() {
...
+      var privkey = new PrivateKey(bn);
+      var buffer = privkey.toBuffer();
+      buffer.length.should.equal(32);
+    });
+
+    it('will output a 31 byte buffer', function() {
+      var bn = BN.fromBuffer(new Buffer('9b5a0e8fee1835e21170ce1431f9b6f19b487e67748ed70d8a4462bc031915', 'hex'));
+      var privkey = new PrivateKey(bn);
+      var buffer = privkey.toBufferNoPadding();
+      buffer.length.should.equal(31);
```
Comment (Levino) [inline test/privatekey.js]: Is this something that one wants? What is the usecase in the future?

File: lib/hdprivatekey.js (hunk 3 above)
Comment (braydonf) [inline lib/hdprivatekey.js]: The nonCompliant derivation (bug) only affects hardened derivations.

File: test/hdkeys.js (hunk 5 above)
Comment (braydonf) [inline test/hdkeys.js]: It's not  related to the zero padding bug, it has to do with behavior if the derived private key is invalid, which is nearly impossible.

File: test/privatekey.js (hunk 6 above)
Comment (braydonf) [inline test/privatekey.js]: For the same reason that `deriveNonCompliant` exists. If there is hashing of privateKey.toBuffer(), the padding will make a difference, and thus there is still a method for it.

File: lib/hdprivatekey.js (hunk 1 above)
Comment (braydonf) [inline lib/hdprivatekey.js]: I feel like the method name `deriveNonCompliantChild` is warning enough.

File: lib/hdprivatekey.js [hunk 7]
```diff
@@ -127,6 +127,11 @@ HDPrivateKey._getDerivationIndexes = function(path) {
   return _.any(indexes, isNaN) ? null : indexes;
 };
 
+HDPrivateKey.prototype.derive = function() {
+  throw new Error('Method has been deprecated, please use deriveChild or deriveNonCompliantChild' +
+                  'and see documentation for more information');
```
Comment (braydonf) [inline lib/hdprivatekey.js]: This is to make sure that if someone upgrades to 1.0.0, that they will not mistakenly be unaware of the bug that affected the deprecated function.

Comment (Levino) [inline lib/hdprivatekey.js]: I unterstand that. So just completely remove the method. This will also throw an error of someone wants to use it, but then it does not show in IDE autocomplete and so on. Just think how annoying this method will be in the future.

File: lib/hdprivatekey.js [hunk 8]
```diff
@@ -139,28 +144,51 @@ HDPrivateKey._getDerivationIndexes = function(path) {
...
  * assert(child_0_1_2h.xprivkey === copy_of_child_0_1_2h);
  * ```
  *
  * @param {string|number} arg
  * @param {boolean?} hardened
+ * @param {boolean?} nonCompliant - This should not be used and only for backwards compatibility
+ *
  */
-HDPrivateKey.prototype.derive = function(arg, hardened) {
+HDPrivateKey.prototype.deriveChild = function(arg, hardened, nonCompliant) {
```
Comment (Levino) [inline lib/hdprivatekey.js]: This is the most important remark of my review.

**Comment by matiu** [comment]
ACK. Lets merge this and move forward.

**Comment by isocolsky** [comment]
ACK

File: lib/hdprivatekey.js (hunk 8 above)
Comment (gabegattis) [inline lib/hdprivatekey.js]: Does the implementation in https://github.com/bitpay/bitcore-lib/pull/116 satisfy some of your comments? It was based off of this PR, but with some slight changes.

Comment (braydonf) [inline lib/hdprivatekey.js]: I'll take a look at updating, though I agree, nonCompliant likely shouldn't be hanging  around on public API method.

**Comment by braydonf** [comment]
@matiu @isocolsky Sounds good!

**Comment by mpolci** [comment]
Why this PR is not yet merged?

**Comment by matiu** [comment]
This one was replaced by https://github.com/bitpay/bitcore-lib/pull/116
